File: System-Programming-Fundamentals-5th-Term/Lab5-Interprocess-Communication/sysloadsockclt.pl
#!/usr/bin/perl -T

use strict;
use warnings qw(FATAL all);
use IO::Socket::UNIX;

use constant STATE_FMT => 'i<i<i<d<d<d<d<';
use constant STATE_SIZE => length pack(STATE_FMT, 0, 0, 0, 0, 0, 0);
use constant LAB_SOCK_PATH => '/tmp/spf-lab5-sysload-sock';

my $client = IO::Socket::UNIX->new(Type => SOCK_STREAM(), Peer => LAB_SOCK_PATH)
  or die "Failed to connect to a socket: $!\n";

my $state_buf = "";
read $client, $state_buf, STATE_SIZE;
my ($spid, $suid, $sgid, $rt, $sl1, $sl5, $sl15) = unpack STATE_FMT, $state_buf;

print "Connected to server with pid=$spid, uid=$suid, gid=$sgid, runtime=${rt}s\n";
print "Load average:\n  1 minute: $sl1\n  5 minutes: $sl5\n  15 minutes: $sl15\n";

File: System-Programming-Fundamentals-5th-Term/Lab5-Interprocess-Communication/sysloadsockclt.c
#define _DEFAULT_SOURCE
#include <unistd.h>
#include <stdint.h>
#include <stdlib.h>

#include <sys/un.h>
#include <sys/types.h>
#include <sys/socket.h>

#include "error.h"
#include "sysloadsock.h"

int main(unused(int argc), unused(char** argv)) {
  int sockfd;
  CHK_ERRNO(sockfd = socket(AF_UNIX, SOCK_STREAM, 0));

  struct sockaddr_un sockaddr;
  sockaddr.sun_family = AF_UNIX;
  strncpy(sockaddr.sun_path, LAB_SOCK_PATH, sizeof(sockaddr.sun_path) - 1);
  unsigned int sockaddr_len = sizeof(struct sockaddr_un);

  CHK_ERRNO(connect(sockfd, (const struct sockaddr *) &sockaddr, sockaddr_len));

  server_state_t state;
  CHK_ERRNO(read(sockfd, &state, sizeof(server_state_t)));

  printf("Connected to server with pid=%jd, uid=%jd, gid=%jd\n",
      (intmax_t) state.pid, (intmax_t) state.uid, (intmax_t) state.gid);
  printf("Load average:\n  1 minute: %.2lf\n  5 minutes: %.2lf\n  15 minutes: %.2lf\n",
      state.loadavg[0], state.loadavg[1], state.loadavg[2]);

  close(sockfd);

  return 0;
}
